Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28226 Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)**

# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

xush6528 committed:
Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)

# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]

xush6528 committed:
Update on "Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)"


# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]

xush6528 committed:
Update on "Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)"


# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]

xush6528 committed:
Update on "Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)"


# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -6,15 +6,19 @@
+        backend=RpcBackend.PROCESS_GROUP,
+        init_method=None,
+        self_rank=-1,
+        worker_name_to_id=None,
```
It seems the `Store.hpp` API only allows `int64_t` value type. @pietern will it be acceptable if we modify the store API to take string values?

xush6528 replied:
We can't pass `world_size` only, because for c10d::Store, there is no "list" API to list keys of other participant nodes. There is no way to wait until the number participants equals `world_size`.

xush6528 replied:
Or we use the "add" API as an atomic counter? Every one publish it's only address and add the counter?

mrshenli replied:
If this PR is necessary to unblock, let's get this in, but would prefer to not having a `worker_name_to_id` arg in the final API. We can discuss how to modify stores to handle this use case.

satgera replied:
I agree Shen this API is getting cluttered, we need to clean this up a bit.

xush6528 replied:
@satgera 

Made a first step for cleaning this up in #28628.

xush6528 commented on test/dist_utils.py:
```diff
@@ -47,71 +47,77 @@ def on_master_follower_report_done(worker_name):
-        rpc.join_rpc()
 
-    return wrapper
+            old_test_method(self, *arg, **kwargs)
```
For readability, since I added a layer of function. 
Now it becomes, `dist_init `, `decorator`, `new_test_method`.

xush6528 commented on test/dist_utils.py:
```diff
@@ -47,71 +47,77 @@ def on_master_follower_report_done(worker_name):
+                "worker{}".format(rank): rank for rank in range(self.world_size)
+            }
+            global _ALL_NODE_NAMES
+            _ALL_NODE_NAMES = self.worker_name_to_id.keys()
```
Right. I'll me move it.

xush6528 committed:
Update on "Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)"


# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]

xush6528 committed:
Update on "Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)"


# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]

mrshenli commented:
Looking good to me. Test failures are relevant.

```
Oct 22 23:16:09 ======================================================================
Oct 22 23:16:09 ERROR: test_invalid_names (__main__.RpcTestWithFork)
Oct 22 23:16:09 ----------------------------------------------------------------------
Oct 22 23:16:09 Traceback (most recent call last):
Oct 22 23:16:09   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 130, in wrapper
Oct 22 23:16:09     self._join_processes(fn)
Oct 22 23:16:09   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 211, in _join_processes
Oct 22 23:16:09     self._check_return_codes(elapsed_time)
Oct 22 23:16:09   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 230, in _check_return_codes
Oct 22 23:16:09     raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time))
Oct 22 23:16:09 RuntimeError: Process 0 terminated or timed out after 400.4009075164795 seconds
```

xush6528 committed:
Update on "Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)"


# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Stamped to unblock. Please land after test pass.

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -41,8 +41,22 @@ def register_rendezvous_handler(scheme, handler):
 
-def rendezvous(url, **kwargs):
-    global _rendezvous_handlers
+def rendezvous(url, rank=-1, world_size=-1, **kwargs):
```
This is changing the `rendezvous` Python API. @pietern do you know if anyone using this API directly?

xush6528 committed:
Update on "Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)"


# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]

xush6528 commented:
Move code to avoid rendezvous twice.

xush6528 committed:
Update on "Make init_method url appending step re-usable by both init_process_group and init_model_parallel(init_rpc)"


# Goal

Rendezvous step should be the first step not only for `init_process_group` but also for `init_model_parallel`.

The road block is that there is special step in `init_process_group` where arguments `rank`, `world_size` passed to `init_process_group(..)` are appended to `init_method` url string.

We need to make this argument appending step common and re-usable for both `init_process_group` and `init_model_parallel`.

# Solution

- Put argument appending inside of `rendezvous` function.
- Remove manual `init_method` url construction. Delegate the responsibility to the `rendezvous` function.
- Use the `rendezvous` function for any `RpcAgent`.

Differential Revision: [D5524494](https://our.internmc.facebook.com/intern/diff/D5524494/)

[ghstack-poisoned]
